packages/core/src/tools/sanitize.ts

```diff
@@ -0,0 +1,235 @@
+ * initial call to sanitize, and iterations for Arrays and Objects
+ *
+ */
+function sanitizeProcessor(
+  source: ExtendedContextValue,
+  parentPath: string,
+  key: string | number | undefined,
+  queue: ContainerElementsToProcess[],
+  visitedObjectsWithPath: WeakMap<object, string>
+) {
+  // Start by handling toJSON, as we want to sanitize its output
+  const sourceToSanitize = tryToApplyToJSON(source)
+
+  if (!sourceToSanitize || typeof sourceToSanitize !== 'object') {
+    return sanitizePrimitivesAndFunctions(sourceToSanitize)
+  }
+
+  const sanitizedSource = sanitizeObjects(sourceToSanitize)
+  if (sanitizedSource !== '[Object]' && sanitizedSource !== '[Array]') {
+    return sanitizedSource
+  }
+
+  // Handle potential cyclic references
+  // We need to use source as sourceToSanitize could be a reference to a new object
+  // At this stage, we know the source is an object type
+  const sourceAsObject = source as object
+  if (visitedObjectsWithPath.has(sourceAsObject)) {
+    return `[Reference seen at ${visitedObjectsWithPath.get(sourceAsObject)!}]`
+  }
+
+  // Add processed source to queue
+  const currentPath = key ? `${parentPath}.${key}` : `${parentPath}`
+  const target = Array.isArray(sourceToSanitize) ? [] : {}
+  visitedObjectsWithPath.set(sourceAsObject, currentPath)
+  queue.push({ source: sourceToSanitize, target, path: currentPath })
+
+  return target
+}
+
+/**
+ * Used to approximate the number of characters elements occupy inside a container
+ * "key":"value" => returns a size of 13 if first value in collection (separatorLength=0), 14 otherwise (separatorLength=1)
+ *
+ * LIMITATIONS
+ * - Actual byte count may differ according to character encoding
+ *
+ * @param value              value in key/value pair
+ * @param separatorLength    0 if first pair in container, 1 otherwise (accounts for the comma separator)
+ * @param key                key in key/value pair (undefined for arrays)
+ *
+ */
+function computeSize(value: ContextValue, separatorLength = 0, key?: string) {
+  try {
+    return JSON.stringify(value).length + separatorLength + (key ? key.length + 3 : 0)
+  } catch {
+    // We do not expect this to happen as complex types will have been processed previously
+  }
+  return 0
+}
+
+/**
+ * Handles sanitization of simple, non-object types
+ *
+ */
+function sanitizePrimitivesAndFunctions(value: ExtendedContextValue) {
+  // BigInt cannot be serialized by JSON.stringify(), convert it to a string representation
+  if (typeof value === 'bigint') {
+    return `[BigInt] ${value.toString()}`
+  }
+  // Functions cannot be serialized by JSON.stringify(). Moreover, if a faulty toJSON is present, it needs to be converted
+  // so it won't prevent stringify from serializing later
+  if (typeof value === 'function') {
+    return `[Function] ${value.name || 'unknown'}`
+  }
+  // JSON.stringify() does not serialize symbols. We cannot use (yet) symbol.description as it is part of ES2019+
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏Is it TypeScript forbiding to use `symbol.description`? I would not wait for a TS target upgrade and just cast the value, like `(symbol as SymbolWithDescription).description`. We do that in various places already

Reply:
👌 + added a test